Build failures on several macOS versions: `analysis_xtensa.c: error: too many arguments provided to function-like macro invocation`

`rizin` build fails on a few older macOS versions at the moment. There are two issues:

1. On macOS 10.12 and below this fails, at least with Clang (cannot test with GCC right now):
```
FAILED: librz/analysis/librz_analysis.0.7.dylib.p/p_analysis_xtensa.c.o 
/usr/bin/clang -Ilibrz/analysis/librz_analysis.0.7.dylib.p -I. -I../rizin-v0.7.2 -Ilibrz -I../rizin-v0.7.2/librz -Ilibrz/include -I../rizin-v0.7.2/librz/include -I../rizin-v0.7.2/librz/analysis/arch/gb -Ilibrz/util/sdb/src -I../rizin-v0.7.2/librz/util/sdb/src -I../rizin-v0.7.2/librz/asm/arch/include -I../rizin-v0.7.2/librz/asm/arch -I../rizin-v0.7.2/librz/asm/arch/h8300 -I../rizin-v0.7.2/librz/asm/arch/hexagon -I../rizin-v0.7.2/librz/asm/arch/msp430 -I../rizin-v0.7.2/librz/asm/arch/rsp -I../rizin-v0.7.2/librz/asm/arch/mcore -I../rizin-v0.7.2/librz/asm/arch/v850 -I../rizin-v0.7.2/librz/asm/arch/propeller -I../rizin-v0.7.2/librz/asm/arch/ebc -I../rizin-v0.7.2/librz/asm/arch/cr16 -I../rizin-v0.7.2/librz/asm/arch/8051 -I../rizin-v0.7.2/librz/asm/arch/v810 -I../rizin-v0.7.2/librz/asm/arch/or1k -I../rizin-v0.7.2/librz/asm/arch/tricore -I../rizin-v0.7.2/librz/bin/format -I../rizin-v0.7.2/librz/type/parser -I/opt/local/include/capstone -I/opt/local/libexec/openssl3/include -fdiagnostics-color=always -Wall -Winvalid-pch -O2 -g -DRZ_PLUGIN_INCORE=1 -DUSE_SYS_CAPSTONE --std=gnu99 -Werror=sizeof-pointer-memaccess -fvisibility=hidden -pipe -Os -arch x86_64 -MD -MQ librz/analysis/librz_analysis.0.7.dylib.p/p_analysis_xtensa.c.o -MF librz/analysis/librz_analysis.0.7.dylib.p/p_analysis_xtensa.c.o.d -o librz/analysis/librz_analysis.0.7.dylib.p/p_analysis_xtensa.c.o -c ../rizin-v0.7.2/librz/analysis/p/analysis_xtensa.c
../rizin-v0.7.2/librz/analysis/p/analysis_xtensa.c:24:45: error: too many arguments provided to function-like macro invocation
        memcpy(ctx->length_table, (int[16]){ 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 8, 8 }, sizeof(ctx->length_table));
                                                   ^
/usr/include/secure/_string.h:64:9: note: macro 'memcpy' defined here
#define memcpy(dest, src, len)                                  \
        ^
../rizin-v0.7.2/librz/analysis/p/analysis_xtensa.c:24:2: note: parentheses are required around macro argument containing braced initializer list
        memcpy(ctx->length_table, (int[16]){ 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 8, 8 }, sizeof(ctx->length_table));
        ^
                                  (                                                          )
1 error generated.
```
Log: https://build.macports.org/builders/ports-10.12_x86_64-builder/builds/263721/steps/install-port/logs/stdio

2. On 10.6 and below this fails (this is an SDK-related issue, not compiler-related):
```
FAILED: librz/io/librz_io.0.7.dylib.p/p_io_debug.c.o 
/opt/local/bin/clang-mp-11 -Ilibrz/io/librz_io.0.7.dylib.p -I. -I../rizin-v0.7.2 -Ilibrz -I../rizin-v0.7.2/librz -Ilibrz/include -I../rizin-v0.7.2/librz/include -I../rizin-v0.7.2/subprojects/rzgdb/include -I../rizin-v0.7.2/subprojects/rzgdb/include/gdbclient -I../rizin-v0.7.2/subprojects/rzgdb/include/gdbserver -Isubprojects/rzwinkd -I../rizin-v0.7.2/subprojects/rzwinkd -Isubprojects/rzar -I../rizin-v0.7.2/subprojects/rzar -Ilibrz/util/sdb/src -I../rizin-v0.7.2/librz/util/sdb/src -I../rizin-v0.7.2/subprojects/rzqnx/include -I/opt/local/libexec/openssl3/include -fdiagnostics-color=always -Wall -Winvalid-pch -O2 -g -DRZ_PLUGIN_INCORE=1 -DUSE_SYS_CAPSTONE --std=gnu99 -Werror=sizeof-pointer-memaccess -fvisibility=hidden -pipe -Os -arch x86_64 -MD -MQ librz/io/librz_io.0.7.dylib.p/p_io_debug.c.o -MF librz/io/librz_io.0.7.dylib.p/p_io_debug.c.o.d -o librz/io/librz_io.0.7.dylib.p/p_io_debug.c.o -c ../rizin-v0.7.2/librz/io/p/io_debug.c
../rizin-v0.7.2/librz/io/p/io_debug.c:302:2: warning: implicit declaration of function 'posix_spawn_file_actions_addinherit_np' is invalid in C99 [-Wimplicit-function-declaration]
        posix_spawn_file_actions_addinherit_np(&fileActions, STDIN_FILENO);
        ^
../rizin-v0.7.2/librz/io/p/io_debug.c:306:14: error: use of undeclared identifier 'POSIX_SPAWN_CLOEXEC_DEFAULT'
        ps_flags |= POSIX_SPAWN_CLOEXEC_DEFAULT;
                    ^
1 warning and 1 error generated.
```
Log: https://build.macports.org/builders/ports-10.6_x86_64-builder/builds/192137/steps/install-port/logs/stdio

P. S. `POSIX_SPAWN_CLOEXEC_DEFAULT` has been added in 10.7, but apparently is safe to use from 10.8 up. See: https://github.com/python/cpython/issues/109154

Could these be addressed?

GCC gives a better error output:
```
FAILED: librz/analysis/librz_analysis.0.7.dylib.p/p_analysis_xtensa.c.o 
/opt/local/bin/gcc-mp-13 -Ilibrz/analysis/librz_analysis.0.7.dylib.p -I. -I../rizin-v0.7.2 -Ilibrz -I../rizin-v0.7.2/librz -Ilibrz/include -I../rizin-v0.7.2/librz/include -I../rizin-v0.7.2/librz/analysis/arch/gb -Ilibrz/util/sdb/src -I../rizin-v0.7.2/librz/util/sdb/src -I../rizin-v0.7.2/librz/asm/arch/include -I../rizin-v0.7.2/librz/asm/arch -I../rizin-v0.7.2/librz/asm/arch/h8300 -I../rizin-v0.7.2/librz/asm/arch/hexagon -I../rizin-v0.7.2/librz/asm/arch/msp430 -I../rizin-v0.7.2/librz/asm/arch/rsp -I../rizin-v0.7.2/librz/asm/arch/mcore -I../rizin-v0.7.2/librz/asm/arch/v850 -I../rizin-v0.7.2/librz/asm/arch/propeller -I../rizin-v0.7.2/librz/asm/arch/ebc -I../rizin-v0.7.2/librz/asm/arch/cr16 -I../rizin-v0.7.2/librz/asm/arch/8051 -I../rizin-v0.7.2/librz/asm/arch/v810 -I../rizin-v0.7.2/librz/asm/arch/or1k -I../rizin-v0.7.2/librz/asm/arch/tricore -I../rizin-v0.7.2/librz/bin/format -I../rizin-v0.7.2/librz/type/parser -I/opt/local/include/capstone -I/opt/local/libexec/openssl3/include -fdiagnostics-color=always -Wall -Winvalid-pch -O2 -g -Wimplicit-fallthrough=3 -DRZ_PLUGIN_INCORE=1 -DUSE_SYS_CAPSTONE --std=gnu99 -Werror=sizeof-pointer-memaccess -fvisibility=hidden -pipe -Os -arch ppc -MD -MQ librz/analysis/librz_analysis.0.7.dylib.p/p_analysis_xtensa.c.o -MF librz/analysis/librz_analysis.0.7.dylib.p/p_analysis_xtensa.c.o.d -o librz/analysis/librz_analysis.0.7.dylib.p/p_analysis_xtensa.c.o -c ../rizin-v0.7.2/librz/analysis/p/analysis_xtensa.c
../rizin-v0.7.2/librz/analysis/p/analysis_xtensa.c: In function 'xtensa_init':
../rizin-v0.7.2/librz/analysis/p/analysis_xtensa.c:24:121: error: macro "memcpy" passed 18 arguments, but takes just 3
   24 |         memcpy(ctx->length_table, (int[16]){ 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 8, 8 }, sizeof(ctx->length_table));
      |                                                                                                                         ^
In file included from /usr/include/string.h:148,
                 from ../rizin-v0.7.2/librz/analysis/p/analysis_xtensa.c:4:
/usr/include/secure/_string.h:50: note: macro "memcpy" defined here
   50 | #define memcpy(dest, src, len)                                  \
      | 
../rizin-v0.7.2/librz/analysis/p/analysis_xtensa.c:24:9: warning: statement with no effect [-Wunused-value]
   24 |         memcpy(ctx->length_table, (int[16]){ 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 8, 8 }, sizeof(ctx->length_table));
      |         ^~~~~~
```

--- a/librz/bin/dwarf/endian_reader.h
+++ b/librz/bin/dwarf/endian_reader.h
@@ -140,7 +140,7 @@ static inline bool R_read_cstring(RzBinEndianReader *R, const char **x) {
 	if (!(R->data && R->offset + 1 <= R->length)) {
 		return false;
 	}
-	ut64 len = strnlen((char *)R_data(R), R_remain(R));
+	ut64 len = rz_str_nlen((char *)R_data(R), R_remain(R));
 	*x = (const char *)R_data(R);
 	R->offset += len + 1;
 	return true;

--- a/librz/debug/p/native/xnu/xnu_debug.c
+++ b/librz/debug/p/native/xnu/xnu_debug.c
@@ -500,7 +500,11 @@ RzList *xnu_thread_list(RzDebug *dbg, int pid, RzList *list) {
 #if __arm__ || __arm64__ || __aarch_64__
 #define CPU_PC (dbg->bits == RZ_SYS_BITS_64) ? state.arm64.__pc : state.arm32.__pc
 #elif __POWERPC__
+#if __DARWIN_UNIX03
+#define CPU_PC state.__srr0
+#else
 #define CPU_PC state.srr0
+#endif
 #elif __x86_64__ || __i386__
 #define CPU_PC (dbg->bits == RZ_SYS_BITS_64) ? state.uts.ts64.__rip : state.uts.ts32.__eip
 #endif

--- a/librz/debug/p/native/xnu/xnu_debug.c
+++ b/librz/debug/p/native/xnu/xnu_debug.c
@@ -1154,8 +1154,10 @@ static void xnu_map_free(RzDebugMap *map) {
 }
 
 static RzList *xnu_dbg_modules(RzDebug *dbg) {
-#if __POWERPC__
+#if !defined(MAC_OS_X_VERSION_10_7)
 #warning TODO: xnu_dbg_modules not supported
+	// TASK_DYLD_INFO introduced in 10.6
+	// TASK_DYLD_ALL_IMAGE_INFO_* introduced in 10.7
 	return NULL;
 #else
 	rz_return_val_if_fail(dbg && dbg->plugin_data, NULL);

--- a/librz/debug/p/native/xnu/xnu_debug.h
+++ b/librz/debug/p/native/xnu/xnu_debug.h
@@ -72,12 +72,10 @@ int ptrace(int _request, pid_t _pid, caddr_t _addr, int _data);
 #include <sys/fcntl.h>
 #include <sys/proc.h>
 
-// G3
 #if __POWERPC__
 #include <sys/ptrace.h>
 #include <sys/types.h>
 #include <sys/wait.h>
-#include <mach/ppc/_types.h>
 #include <mach/ppc/thread_status.h>
 // iPhone5
 #elif __aarch64
@@ -88,7 +86,6 @@ int ptrace(int _request, pid_t _pid, caddr_t _addr, int _data);
 #elif __arm64
 #include <mach/arm/thread_status.h>
 #else
-// iMac
 /* x86 32/64 */
 #include <mach/i386/thread_status.h>
 #include <sys/ucontext.h>

--- a/librz/debug/p/native/xnu/xnu_excthreads.c
+++ b/librz/debug/p/native/xnu/xnu_excthreads.c
@@ -45,9 +45,8 @@ RZ_IPI bool xnu_modify_trace_bit(RzDebug *dbg, xnu_thread_t *th, int enable) {
 }
 
 #elif __POWERPC__ // ppc processor
-// XXX poor support at this stage i don't care so much. Once intel and arm done it could be done
-// TODO add better support for ppc
-RZ_IPI bool xnu_modify_trace_bit(RzDebug *dbg, void *th, int enable) {
+// TODO: Implement and test this for ppc too. Below is an old example for reference.
+RZ_IPI bool xnu_modify_trace_bit(RzDebug *dbg, xnu_thread_t *th, int enable) {
 	return false;
 }
 #if 0

--- a/librz/debug/p/native/xnu/xnu_threads.c
+++ b/librz/debug/p/native/xnu/xnu_threads.c
@@ -109,7 +109,7 @@ RZ_IPI bool rz_xnu_thread_set_drx(RzXnuDebug *ctx, xnu_thread_t *thread) {
 #ifndef PPC_DEBUG_STATE32
 #define PPC_DEBUG_STATE32 1
 #endif
-	ppc_debug_state_t *regs;
+	// ppc_debug_state_t *regs;
 	// thread->flavor = PPC_DEBUG_STATE32;
 	// thread->count  = RZ_MIN (thread->count, sizeof (regs->uds.ds32));
 	return false;

--- a/librz/debug/p/native/xnu/xnu_threads.c
+++ b/librz/debug/p/native/xnu/xnu_threads.c
@@ -220,7 +220,9 @@ RZ_IPI bool rz_xnu_thread_get_gpr(RzXnuDebug *ctx, xnu_thread_t *thread) {
 }
 
 static bool xnu_fill_info_thread(RzDebug *dbg, xnu_thread_t *thread) {
-#if __POWERPC__
+#if !defined(MAC_OS_X_VERSION_10_6)
+	// THREAD_IDENTIFIER_INFO introduced in 10.6
+	// TODO: use e.g. only basic info as a fallback here
 	thread->name = strdup("unknown");
 	return false;
 #else

--- a/librz/io/p/io_debug.c
+++ b/librz/io/p/io_debug.c
@@ -24,6 +24,13 @@
 #if DEBUGGER && DEBUGGER_SUPPORTED
 #define MAGIC_EXIT 123
 
+// POSIX_SPAWN_CLOEXEC_DEFAULT is available since macOS 10.7, but known to cause kernel panics until 10.8.
+#if __APPLE__ && defined(MAC_OS_X_VERSION_10_8)
+#define USE_POSIX_SPAWN 1
+#else
+#define USE_POSIX_SPAWN 0
+#endif
+
 #include <signal.h>
 #if __UNIX__
 #include <sys/ptrace.h>
@@ -31,10 +38,10 @@
 #include <sys/wait.h>
 #endif
 
-#if __APPLE__
-#if !__POWERPC__
+#if USE_POSIX_SPAWN
 #include <spawn.h>
 #endif
+#if __APPLE__
 #include <sys/types.h>
 #include <sys/wait.h>
 #include <mach/exception_types.h>
@@ -176,56 +183,6 @@ static int fork_and_ptraceme(RzIO *io, int bits, const char *cmd) {
 }
 #else // windows
 
-#if (__APPLE__ && __POWERPC__) || !__APPLE__
-
-#if __APPLE__ || __BSD__
-static void inferior_abort_handler(int pid) {
-	eprintf("Inferior received signal SIGABRT. Executing BKPT.\n");
-}
-#endif
-
-static void trace_me(void) {
-#if __APPLE__
-	rz_sys_signal(SIGTRAP, SIG_IGN); // NEED BY STEP
-#endif
-#if __APPLE__ || __BSD__
-	/* we can probably remove this #if..as long as PT_TRACE_ME is redefined for OSX in rz_debug.h */
-	rz_sys_signal(SIGABRT, inferior_abort_handler);
-	if (ptrace(PT_TRACE_ME, 0, 0, 0) != 0) {
-		rz_sys_perror("ptrace-traceme");
-	}
-#if __APPLE__
-	ptrace(PT_SIGEXC, getpid(), NULL, 0);
-#endif
-#else
-	if (ptrace(PTRACE_TRACEME, 0, NULL, NULL) != 0) {
-		rz_sys_perror("ptrace-traceme");
-		exit(MAGIC_EXIT);
-	}
-#endif
-}
-#endif
-
-#if __APPLE__ && !__POWERPC__
-static void handle_posix_error(int err) {
-	switch (err) {
-	case 0:
-		// eprintf ("Success\n");
-		break;
-	case 22:
-		eprintf("posix_spawnp: Invalid argument\n");
-		break;
-	case 86:
-		eprintf("Unsupported architecture. Please specify -b 32\n");
-		break;
-	default:
-		eprintf("posix_spawnp: unknown error %d\n", err);
-		perror("posix_spawnp");
-		break;
-	}
-}
-#endif
-
 static RzRunProfile *_get_run_profile(RzIO *io, int bits, char **argv) {
 	int i;
 	RzRunProfile *rp = rz_run_new(NULL);
@@ -267,7 +224,24 @@ static RzRunProfile *_get_run_profile(RzIO *io, int bits, char **argv) {
 	return rp;
 }
 
-#if __APPLE__ && !__POWERPC__
+#if USE_POSIX_SPAWN
+
+static void handle_posix_error(int err) {
+	switch (err) {
+	case 0:
+		// 0 means success
+		break;
+	case 22:
+		RZ_LOG_ERROR("posix_spawnp: Invalid argument\n");
+		break;
+	case 86:
+		RZ_LOG_ERROR("Unsupported architecture. Please specify -b 32\n");
+		break;
+	default:
+		RZ_LOG_ERROR("posix_spawnp: unknown error %d\n", err);
+		break;
+	}
+}
 
 static void handle_posix_redirection(RzRunProfile *rp, posix_spawn_file_actions_t *fileActions) {
 	const int mode = S_IRUSR | S_IWUSR;
@@ -282,8 +256,7 @@ static void handle_posix_redirection(RzRunProfile *rp, posix_spawn_file_actions_
 	}
 }
 
-// __UNIX__ (not windows)
-static int fork_and_ptraceme_for_mac(RzIO *io, int bits, const char *cmd) {
+static int fork_and_ptraceme_for_unix(RzIO *io, int bits, const char *cmd) {
 	pid_t p = -1;
 	char **argv;
 	posix_spawn_file_actions_t fileActions;
@@ -340,14 +313,41 @@ static int fork_and_ptraceme_for_mac(RzIO *io, int bits, const char *cmd) {
 	posix_spawn_file_actions_destroy(&fileActions);
 	return p; // -1 ?
 }
-#endif // __APPLE__ && !__POWERPC__
 
-#if (!(__APPLE__ && !__POWERPC__))
+#else // USE_POSIX_SPAWN
+
 typedef struct fork_child_data_t {
 	RzIO *io;
 	int bits;
 	const char *cmd;
 } fork_child_data;
+
+#if __APPLE__ || __BSD__
+static void inferior_abort_handler(int pid) {
+	eprintf("Inferior received signal SIGABRT. Executing BKPT.\n");
+}
+#endif
+
+static void trace_me(void) {
+#if __APPLE__
+	rz_sys_signal(SIGTRAP, SIG_IGN); // NEED BY STEP
+#endif
+#if __APPLE__ || __BSD__
+	/* we can probably remove this #if..as long as PT_TRACE_ME is redefined for OSX in rz_debug.h */
+	rz_sys_signal(SIGABRT, inferior_abort_handler);
+	if (ptrace(PT_TRACE_ME, 0, 0, 0) != 0) {
+		rz_sys_perror("ptrace-traceme");
+	}
+#if __APPLE__
+	ptrace(PT_SIGEXC, getpid(), NULL, 0);
+#endif
+#else
+	if (ptrace(PTRACE_TRACEME, 0, NULL, NULL) != 0) {
+		rz_sys_perror("ptrace-traceme");
+		exit(MAGIC_EXIT);
+	}
+#endif
+}
 
 static void fork_child_callback(void *user) {
 	fork_child_data *data = user;
@@ -408,20 +408,18 @@ static int fork_and_ptraceme_for_unix(RzIO *io, int bits, const char *cmd) {
 	}
 	return child_pid;
 }
-#endif
+
+#endif // USE_POSIX_SPAWN
 
 static int fork_and_ptraceme(RzIO *io, int bits, const char *cmd) {
 	// Before calling the platform implementation, append arguments to the command if they have been provided
-	char *_eff_cmd = io->args ? rz_str_appendf(strdup(cmd), " %s", io->args) : strdup(cmd);
+	char *eff_cmd = io->args ? rz_str_appendf(strdup(cmd), " %s", io->args) : strdup(cmd);
+	if (!eff_cmd) {
+		return -1;
+	}
 	int r = 0;
-
-#if __APPLE__ && !__POWERPC__
-	r = fork_and_ptraceme_for_mac(io, bits, _eff_cmd);
-#else
-	r = fork_and_ptraceme_for_unix(io, bits, _eff_cmd);
-#endif
-
-	free(_eff_cmd);
+	r = fork_and_ptraceme_for_unix(io, bits, eff_cmd);
+	free(eff_cmd);
 	return r;
 }
 #endif

